flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnable.java

```diff
@@ -0,0 +1,252 @@
+
+		Preconditions.checkState(hasAckState || !hasLocalState,
+			"Found cached state but no corresponding primary state is reported to the job " +
+				"manager. This indicates a problem.");
+
+		// we signal stateless tasks by reporting null, so that there are no attempts to assign empty state
+		// to stateless tasks on restore. This enables simple job modifications that only concern
+		// stateless without the need to assign them uids to match their (always empty) states.
+		taskEnvironment.getTaskStateManager().reportTaskStateSnapshots(
+			checkpointMetaData,
+			checkpointMetrics,
+			hasAckState ? acknowledgedTaskStateSnapshot : null,
+			hasLocalState ? localTaskStateSnapshot : null);
+
+		StreamTask.LOG.debug("{} - finished asynchronous part of checkpoint {}. Asynchronous duration: {} ms",
+			taskName, checkpointMetaData.getCheckpointId(), asyncDurationMillis);
+
+		StreamTask.LOG.trace("{} - reported the following states in snapshot for checkpoint {}: {}.",
+			taskName, checkpointMetaData.getCheckpointId(), acknowledgedTaskStateSnapshot);
+	}
+
+	private void handleExecutionException(Exception e) {
+
+		boolean didCleanup = false;
+		AsyncCheckpointState currentState = asyncCheckpointState.get();
+
+		while (AsyncCheckpointState.DISCARDED != currentState) {
+
+			if (asyncCheckpointState.compareAndSet(currentState, AsyncCheckpointState.DISCARDED)) {
+
+				didCleanup = true;
+
+				try {
+					cleanup();
+				} catch (Exception cleanupException) {
+					e.addSuppressed(cleanupException);
+				}
+
+				Exception checkpointException = new Exception(
+					"Could not materialize checkpoint " + checkpointMetaData.getCheckpointId() + " for operator " +
+						taskName + '.',
+					e);
+
+				// We only report the exception for the original cause of fail and cleanup.
+				// Otherwise this followup exception could race the original exception in failing the task.
+				try {
+					taskEnvironment.declineCheckpoint(checkpointMetaData.getCheckpointId(), checkpointException);
+				} catch (Exception unhandled) {
+					AsynchronousException asyncException = new AsynchronousException(unhandled);
+					asyncExceptionHandler.handleAsyncException("Failure in asynchronous checkpoint materialization", asyncException);
+				}
+
+				currentState = AsyncCheckpointState.DISCARDED;
+			} else {
+				currentState = asyncCheckpointState.get();
+			}
+		}
+
+		if (!didCleanup) {
+			StreamTask.LOG.trace("Caught followup exception from a failed checkpoint thread. This can be ignored.", e);
+		}
+	}
+
+	@Override
+	public void close() {
+		if (asyncCheckpointState.compareAndSet(AsyncCheckpointState.RUNNING, AsyncCheckpointState.DISCARDED)) {
+
+			try {
+				cleanup();
+			} catch (Exception cleanupException) {
+				StreamTask.LOG.warn("Could not properly clean up the async checkpoint runnable.", cleanupException);
+			}
+		} else {
+			logFailedCleanupAttempt();
+		}
+	}
+
+	private void cleanup() throws Exception {
+		StreamTask.LOG.debug(
```

Comment:
Shall we pass the logger from outside instead of expose it as `public/(default)`?

Reply:
It's already exposed for the other classes but yes, it's better not to use it.
I'll change it to a private logger.